### ISSUE TITLE
Request for implementation of `Gradient calcGradient(Gradient layerError, INDArray activation)`

### ISSUE THREAD

**Issue by dkmisra** [deeplearning4j/deeplearning4j#899]
## Bug

Function `calcGradient(Gradient layerError, INDArray activation)` throws
`UnsupportedOperationException()` in MultiLayerNetwork.java
## Why is this function needed

For those of us interested in having a finer control over learning or performing operations in computation graph, we want to be able to propagate the gradient (computed using some other complex program or hand-tuned for debugging) through the previous layers of the network.
Java doc of `calcGradient` suggets that this function was designed with this in mind however its currently not supported.

**Comment by AlexDBlack** [comment]
Following up our gitter conversation, this would require two things:
- An implementation of MultiLayerNetwork.backpropGradient(INDArray epsilon)
- Methods to actually use the returned gradients (use updaters, etc) to update the parameters

**Comment by dkmisra** [comment]
I will just post that gist link which tries to emulate the core of this. 

https://gist.github.com/dkmisra/bda3c551cc624a414765

As you mentioned, this only returns raw gradient (no AdaGrad, clipping etc.) and does not perform reshaping etc.

**Comment by AlexDBlack** [comment]
So I had a look into this. As noted, there's two issues here:
First: calculating gradients. This is pretty easy to implement (just modify existing backprop() method to allow this, in addition to 'standard' usage with OutputLayer). I've already got this essentially implemented locally (not on github yet).

Second: actually using the gradients. This is the harder part, given the design for DL4J.
Implementing something to use them isn't so hard, but doing it in a way that makes sense from a design point of view kinda is. We can only really do this for SGD (not other optimizers that use line search) as others all require recalculation of gradients and score. I'm quite hesitant to add an "updateParameters(Gradients)" method or equivalent that works *only* for SGD.
What I'm thinking is that in the short term, the best solution is to apply them manually. For example: https://gist.github.com/AlexDBlack/609099bcaf6ce90f805c
Longer term, we focus on getting the computation graph properly implemented to do this properly (which is on our roadmap for some point in the future anyway)

**Comment by AlexDBlack** [comment]
I've implemented MultiLayerNetwork.backpropGradient(INDArray epsilon), which is the first step above.
This basically allows us to use the MultiLayerNetwork as we would use any other layer (from a design point of view, MultiLayerNetwork *is* a Layer)
https://github.com/deeplearning4j/deeplearning4j/pull/901

See also https://github.com/deeplearning4j/deeplearning4j/blob/master/deeplearning4j-core/src/test/java/org/deeplearning4j/nn/multilayer/MultiLayerTest.java#L479 for one way to update parameters/use gradients returned by this method.

Is that enough for your intended use?

**Comment by dkmisra** [comment]
Just to clarify, this image describes what I need (along with the notations I use below). https://docs.google.com/drawings/d/1yXx6ASgPE9f7LLSNg6Qxt2-qXyPU_-Z_kiWKFh-bM1U/edit?usp=sharing

I don't think your code returns `del loss / del x` which is needed for computation graph. That said, I don't think it should be difficult. 

I was not able to use your code. Not sure if its part of 3.7 since it still gives me *"Not yet implemented" * exception. I also have few questions regarding the code you sent:

1. > `Pair<Gradient,INDArray> pair = net.backpropGradient(eps);`

     I believe this returns `del loss /del w` for every layer. Right?

2. > ```//Apply learning rate, gradient clipping, adagrad/momentum/rmsprop etc
   > Updater updater = UpdaterCreator.getUpdater(net);
   > updater.update(net, g, 0, miniBatch);```

   By that comment you mean that Updater already does that right?

3.  > ```INDArray params = net.params();
     > stepFunction.step(params, g.gradient());
     > net.setParams(params);    //params() may not be in-place```

   What is this step doing? 

Thanks. Let me know how I can test the code. Returning `del loss/del x`  is really crucial to my work and in general for any computation graph purpose.

**Comment by AlexDBlack** [comment]
"I was not able to use your code. Not sure if its part of 3.7 since it still gives me *"Not yet implemented" * exception"
Yes, you'll need to be on DL4J master (github) to use it, until the next release.

Re: your questions:

1. Pair<Gradient,INDArray> is two things:
(a) Gradient object, which is basically a Map<String,INDArray>
Keys: type of parameter. In this case: 0_W would be weights of 0th layer; 1_b would be biases for 1st layer, etc.
Values: the gradients, as an INDArray. dLoss/dParameter
(b) INDArray, which we are referring to as epsilon. See below.


2. Yes, the updater applies learning rate, momentum, etc. It has internal state for most updaters (momentum, rmsprop etc) hence you want to re-use it
In normal learning, the updater is stored in the MultiLayerNetwork and re-used automatically.


3. That's actually updating the parameters given the (updated / LR-applied) gradients.
Specifically, net.params() gives us a 1xN vector of parameters (N params total in network), and Gradient.gradient() flattens the gradients to 1xN in the exact same order.
stepFunction.step(params,g.gradients()) is literally just params -= gradients. (For negative default step function, anyway)


Epsilon / INDArray in Pair<Gradient,INDArray>
So we need the error (delta) of a layer in terms of the next layer
http://neuralnetworksanddeeplearning.com/chap2.html eq. BP2 for example for a standard MLP (RNNs like LSTM have a different form of this equation that includes a sum)
from a design point of view, this equation (BP2) is hard to implement directly without ending up with weird/bad design dependencies across layers
So, instead: return epsilon (i.e., epsilon = delta *before* doing element-wise multiply by sigmaPrime(z) in BP2)
https://github.com/deeplearning4j/deeplearning4j/blob/master/deeplearning4j-core/src/main/java/org/deeplearning4j/nn/layers/BaseLayer.java#L143

**Comment by agibsonccc** [comment]
https://github.com/deeplearning4j/dl4j-examples/blob/master/dl4j-examples/src/main/java/org/deeplearning4j/examples/misc/externalerrors/MultiLayerNetworkExternalErrors.java
